app/controllers/amr_heatmap_controller.rb

```diff
@@ -68,169 +69,67 @@ def amr_counts
-    end
-    return card_owl
+  def set_aws_client
+    @client = Aws::S3::Client.new(region: DEFAULT_S3_REGION)
   end
 
-  def search_card_owl(gene_name, xml_dom)
-    search_result = {}
-    err = "No match found for %s in the CARD Antibiotic Resistance Ontology."
-    owl_classes = xml_dom.xpath(".//owl:Class")
-    matching_entry = nil
-
-    # normalize gene name
-    alphanumeric_gene_name = gene_name.downcase.gsub(/\W/, '')
-    regex_for_gene = Regexp.new("\\b#{alphanumeric_gene_name}\\b")
-    # search through owl classes
-    owl_classes.each do |owl_class|
-      # first we check the label
-      label = owl_class.at_xpath("./rdfs:label").content
-      separated_label = label.downcase.tr('/', ' ')
-      alphanumeric_label = separated_label.gsub(/[^0-9a-z_\s]/, '')
-      if regex_for_gene.match?(alphanumeric_label)
-        matching_entry = owl_class
-        break
-      end
-
-      # then we check the synonyms
-      synonyms = owl_class.xpath("./oboInOwl:hasExactSynonym")
-      synonyms.each do |synonym|
-        separated_synonym = synonym.content.downcase.tr('/', ' ')
-        alphanumeric_synonym = separated_synonym.gsub(/[^0-9a-z_\s]/, '')
-        if regex_for_gene.match?(alphanumeric_synonym)
-          matching_entry = owl_class
-          break
-        end
-      end
-    end
-
-    # abort if no matching entry
-    if matching_entry.nil?
-      search_result["error"] = format(err, gene_name)
-    else
-      search_result["matching_node"] = matching_entry
-    end
-
-    return search_result
+  def fetch_current_ontology_file_key
+    ontology_folder = @client.list_objects_v2(bucket: S3_JSON_BUCKET,
+                                              prefix: S3_JSON_PREFIX).to_h
+    target_key = ontology_folder[:contents][-1][:key] # contents already sorted by key
+    return target_key
   end
 
-  def gather_ontology_information(matching_entry, xml_dom)
-    ontology_information = {}
-    # Gathering properties
-    ontology_information["label"] = matching_entry.at_xpath("./rdfs:label").content
-    ontology_information["accession"] = matching_entry.at_xpath("./oboInOwl:id").content.split(':')[1]
-    ontology_information["description"] = matching_entry.at_xpath("./obo:IAO_0000115").content
-
-    # These properties are referred to by accession number in the matching OWL Class,
-    # but their labels are located elsewhere in the document.
-    remote_properties = {
-      "drugClass" => get_drug_classes(ontology_information["accession"], xml_dom),
-      "geneFamily" => matching_entry.xpath("./rdfs:subClassOf[not(*)]"),
+  def fetch_ontology(s3_key, gene_name)
+    s3_select_params = {
+      bucket: S3_JSON_BUCKET,
+      key: s3_key,
+      expression_type: "SQL",
+      expression: "SELECT * FROM S3Object[*].#{gene_name} LIMIT 1",
+      input_serialization: {
+        json: {
+          type: "DOCUMENT",
+        },
+      },
+      output_serialization: {
+        json: {},
+      },
```

Comment:
Awesome use of the S3 SDK! I can see this being useful in other situations as well. Can you generalize this function and put it in the lib folder?

Reply:
Thanks! It was Charles' idea. I'll see if I can generalize it.